src/qukeys/Qukeys.cpp

```diff
@@ -20,154 +20,197 @@ namespace qukeys {
+  // have been any events that cause it to become either primary or alternate.
+  for (byte i{1}; i < event_queue_.length(); ++i) {
+    // First deal with key press events:
+    if (event_queue_.isPress(i)) {
+      if (qukey_is_spacecadet) {
+        queued_event.key = qukey.primaryKey();
+        event_queue_.remove(0);
+        return true;
+      }
+      if (next_keypress_index == 0) {
+        next_keypress_index = i;
       }
+      continue;
+    }
+    // event_queue_[i] is a release event
+
+    // If this is a release of the qukey
+    if (event_queue_.addr(i) == queued_event.addr) {
+      if (next_keypress_index == 0 || overlap_required_ == 0) {
+        // there were no keypresses between the qukey press and its release, or
+        // there's no release delay overlap configured.
+        queued_event.key = qukey_is_spacecadet ? qukey.alternateKey() : qukey.primaryKey();
+        event_queue_.remove(0);
+        return true;
+      }
+      // calculate release delay and check to see if it has timed out
+      uint16_t overlap_start = event_queue_.timestamp(next_keypress_index);
+      uint16_t overlap_end = event_queue_.timestamp(i);
+      if (releaseDelayed(overlap_start, overlap_end)) {
+        continue;
+      }
+      queued_event.key = qukey.primaryKey();
+      event_queue_.remove(0);
+      return true; 
     }
 
-    // Now we can flush the remaining head of the queue of any non-qukeys:
-    flushQueue();
-
-    // Next, if there was a delayed qukey release, send its release event:
-    if (delayed_qukey_release.addr.isValid()) {
-      controller_.handleKeyEvent(delayed_qukey_release);
+    for (byte j{1}; j < i; ++j) {
+      if (event_queue_.isPress(j) &&
+          event_queue_.addr(j) == event_queue_.addr(i)) {
+        // event_queue_[i] is a release of a key that was pressed subsequent
+        // to the qukey
+        queued_event.key = qukey.alternateKey();
+        event_queue_.remove(0);
+        return true;
+      }
     }
 
-    // Finally, we need to allow the current release event to proceed. But
-    // first, we must correct the key value from what's in the controller's
-    // active keys array:
-    event.key = controller_[event.addr];
+    // A key was released that is not in the queue. If it's not a modifier key
+    // (including layer shifts), we send the release event out of order. Without
+    // this block of code, the binary is 30 bytes smaller, but there's a chance
+    // of delaying a release long enough that it repeats, despite the user
+    // having tapped the key. Also, the queue is more likely to overflow,
+    // requiring a premature release of a qukey.
+    KeyAddr k = event_queue_.addr(i);
+    Key key = controller_[k];
+    if (! (isModifierKey(key) || isLayerShiftKey(key))) {
+      queued_event.addr = k;
+      queued_event.key = key;
+      queued_event.state = cKeyState::release;
+      event_queue_.remove(i);
+      return true;
```

Comment:
Removing this whole section fixes the rollover problem from non-modifiers to qukeys. I'm not sure why, however.

Reply:
The problem was in `EventQueue::remove()`. I failed to update the release state bitfield correctly.